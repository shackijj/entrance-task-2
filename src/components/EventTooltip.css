.EventTooltip {
  position: relative;
  display: inline-block;
  background: #FFFFFF;
  padding: 16px;
  box-shadow: 0 1px 16px 0 rgba(0,44,92,0.28);
  border-radius: 8px;
  min-width: 330px;
}

.EventTooltip-Title {
  font-size: 1.154em;
  line-height: 1.846em;
  font-weight: bold;
}

.EventTooltip-Users {
  margin-top: 16px;
  display: inline-flex;
  align-items: center;
}

.EventTooltip-Button {
  position: absolute;
  top: 8px;
  right: 8px;
}

.EventTooltip-MoreUsers {
  color: rgba(0, 0, 0, 0.5);
}