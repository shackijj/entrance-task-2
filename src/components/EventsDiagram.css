.EventsDiagram {
  overflow-y: hidden;
  overflow-x: scroll;
  position: relative;
}

.EventsDiagram_noscroll {
  overflow-x: hidden;
}

.EventsDiagram-Timeline {
  margin-left: 255px;
}

.EventDiagram-Room {
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  height: 44px;
}

.EventsDiagram-HorizontalScrollContainer {
  display: inline-block;
  height: 100%;
}

.EventsDiagram-VerticalScrollContainer {
  height: calc(100% - 45px);
  overflow-y: scroll;
}

.EventsDiagram_noscroll .EventsDiagram-VerticalScrollContainer {
  overflow-y: hidden;
}

.EventDiagram-FloorTitle {
  margin-top: 21px;
  margin-left: 24px;
  margin-bottom: 4px;
  font-weight: bold;
  text-transform: uppercase;
  color: #858E98;
  font-size: 0.846em;
}

.EventDiagram-Room .Room {
  margin-left: 24px;
  width: 231px;
}

.EventDiagram-Room .RoomTimeline {
  flex: 1;
}

.EventsDiagram-TooltipContainer {
  position: absolute;
  z-index: 9;
  min-width: 330px;
  min-height: 115px;
}