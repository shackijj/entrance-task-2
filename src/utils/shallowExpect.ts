import { compose } from 'redux';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

export default compose(expect, toJSON, shallow);