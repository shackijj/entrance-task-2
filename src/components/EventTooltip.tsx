import * as React from 'react';
import User from './User';
import RoundButton from './RoundButton';
import { Edit } from './GlyphIcon/GlyphIcon';
import * as moment from 'moment';
import './EventTooltip.css';

import { graphql } from 'react-apollo';
import gql from 'graphql-tag';

type User = {
  login: string;
  avatarUrl: string;
};

type Room = {
  title: string;
};

type Event = {
  title: string;
  dateStart: string;
  dateEnd: string;
  room: Room;
  users: User[];
};

interface EventTooltipProps extends Event {
  style?: React.CSSProperties;
}

export const EventTooltip: React.SFC<EventTooltipProps> = ({title, room, dateStart, dateEnd, users, style}) => {
  const momentStart = moment(dateStart);
  const momentEnd = moment(dateEnd);
  const date = momentStart.format('D MMMM');
  const start = momentStart.format('HH:mm');
  const end = momentEnd.format('HH:mm');
  
  let moreUsers;
  const moreUsersLen = users.length - 1;
  switch (moreUsersLen) {
    case 0:
      moreUsers = '';
      break;
    case 1:
      moreUsers = `и еще 1 участник`;
      break;
    case 2:
    case 3:
    case 4:
      moreUsers = `и еще ${moreUsersLen} участника`;
      break;
    default:
      moreUsers = `и еще ${moreUsersLen} участников`;
  } 

  return (
    <div className="EventTooltip" style={style}>
      <RoundButton classes={['EventTooltip-Button']} icon={<Edit/>}/>
      <div className="EventTooltip-Title">{title}</div>
      <div className="EventTooltip-Info">
        {date}, {start}—{end} · {room.title}
      </div>
      <div className="EventTooltip-Users">
        <User {...users[0]}/>
        <span className="EventTooltip-MoreUsers">{moreUsers}</span>
      </div>
    </div>
  );
};

const TOOLTIP_QUERY = gql`
  query EventQuery($id: ID!) {
    event(id: $id) {
      dateStart
      dateEnd
      room {
        title
      }
      users {
        login
        avatarUrl
      }
    }
  }
`;

type Response = {
  event: Event;
};

type InputProps = {
  id: string;
  style?: React.CSSProperties;
};

export const withGraphQL = graphql<Response, InputProps>(TOOLTIP_QUERY, {
  options: ({id}) => ({
    variables: { id }
  })
});

export default withGraphQL(({data, style}) => {
  if (data && data.event) {
    return <EventTooltip {...data.event} style={style}/>;
    
  }
  return <div/>;
});