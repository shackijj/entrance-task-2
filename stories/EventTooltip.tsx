import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { EventTooltip } from '../src/components/EventTooltip';

const event = {
  title: 'Событие 3',
  dateStart: '2018-01-09T18:30:00.55',
  dateEnd: '2018-01-09T20:45:00.55',
  room: {
    title: 'Желтый дом'
  },
  users: [
    {
      login: 'Дарт Вейдер 1',
      avatarUrl: 'https://pp.userapi.com/c402317/v402317531/6107/Tg1PWZHDAO0.jpg'
    },
    {
      login: 'Дарт Вейдер 2',
      avatarUrl: 'https://pp.userapi.com/c402317/v402317531/6107/Tg1PWZHDAO0.jpg'
    },
    {
      login: 'Дарт Вейдер 3',
      avatarUrl: 'https://pp.userapi.com/c402317/v402317531/6107/Tg1PWZHDAO0.jpg'
    }
  ]
};

storiesOf('EventTooltip', module)
  .add('default', () => (
    <div>
      <EventTooltip {...event}/>
    </div>));