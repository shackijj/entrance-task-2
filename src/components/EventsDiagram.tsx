import * as React from 'react';
import Timeline from './Timeline';
import Room from './Room';
import RoomTimeline from './RoomTimeline';
import EventTooltip from './EventTooltip';
import * as classNames from 'classnames';
import { Floor } from './MainPage';
import { HOUR_START, HOUR_END } from '../constants';

import './EventsDiagram.css';

interface EventsDiagramProps {
  floors: Floor[];
  classes: string[];
  dateCurrent?: Date;
}

type Tooltip = {
  id: string;
  style: React.CSSProperties;
};

interface EventsDiagramState {
  tooltip: Tooltip | null;
}

class EventsDiagram extends React.Component<EventsDiagramProps, EventsDiagramState> {
  public onEventClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  private _container: HTMLDivElement | null;
  constructor(props: EventsDiagramProps) {
    super(props);
    this.state = {
      tooltip: null
    };
    this.onEventClick = this._onEventClick.bind(this);
  }

  public render() {
    const {floors, classes, dateCurrent} = this.props;
    const { tooltip }  = this.state;
    return (
      <div className={classNames('EventsDiagram', classes)} ref={div => this._container = div}>
        <div className="EventsDiagram-HorizontalScrollContainer">
          <Timeline 
            classes={['EventsDiagram-Timeline']}
            hourStart={HOUR_START}
            hourEnd={HOUR_END}
            dateCurrent={dateCurrent}
          />

          <div className="EventsDiagram-VerticalScrollContainer">
            {floors.map((floor, floorIdx) => (
              <div className="EventDiagram-Floor" key={floorIdx}>
                <div className="EventDiagram-FloorTitle">{floor.floor} этаж</div>
                {
                  floor.rooms.map((room, roomIdx) => (
                    <div className="EventDiagram-Room" key={roomIdx}>
                      <Room {...room}/>
                      <RoomTimeline
                        dateCurrent={dateCurrent}
                        hourStart={HOUR_START}
                        hourEnd={HOUR_END}
                        onEventClick={this.onEventClick}
                        {...room}
                      />
                    </div>
                  ))
                }
              </div>
            ))}
          </div>
        </div>
        {tooltip && <EventTooltip {...tooltip}/>}
      </div>
    );
  }

  private _onEventClick(event: React.MouseEvent<HTMLDivElement>) {
    if (this._container) {
      const cr = this._container.getBoundingClientRect();
      const er = event.currentTarget.getBoundingClientRect();
      const left = er.left - cr.left + er.width / 2 + this._container.scrollLeft - 362 / 2;
      const lk = cr.width - (left + 362) + this._container.scrollLeft;

      this.setState({
        tooltip: {
          id: '1',
          style: {
            top: er.top - cr.top + er.height + this._container.scrollTop,
            left: lk < 0 ? left + lk : left,
          }
        }
      });
    }
  }
}

export default EventsDiagram;