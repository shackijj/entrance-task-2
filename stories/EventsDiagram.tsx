import * as React from 'react';
import EventsDiagram from '../src/components/EventsDiagram';
import { storiesOf } from '@storybook/react';
import './EventDiagram.css';

const floor = [
  {
    floor: 7,
    rooms: [
      {
        id: '1',
        title: 'Ржавый Фред',
        capacity: 7,
        events: [
          {
            id: '1',
            title: 'Событие 1',
            dateStart: '2018-01-09T08:30:00.55',
            dateEnd: '2018-01-09T08:45:00.55',
            users: [],
          },
        ]
      }
    ]
  }
];

storiesOf('EventsDiagram', module)
  .add('default', () => (
      <EventsDiagram
        floors={floor}
        classes={['EventDiagram-Story']}
        isDateCurrent={false}
        date={'2018-01-09T18:30:00.55'}
      />
  ));
