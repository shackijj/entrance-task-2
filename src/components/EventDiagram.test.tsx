import * as React from 'react';
import EventsDiagram from './EventsDiagram';
import EventTooltip from './EventTooltip';
import shallowExpect from '../utils/shallowExpect';
import { mount } from 'enzyme';

describe('EventDiagram', () => {
  it('should render a floors list', () => {
    const floor = [
      {
        floor: 7,
        rooms: [
          {
            title: 'Ржавый Фред',
            capacity: 7,
            events: [
              {
                title: 'Событие 1',
                dateStart: '2018-01-09T06:30:00.55',
                dateEnd: '2018-01-09T08:45:00.55',
                users: [],
              },
              {
                title: 'Событие 2',
                dateStart: '2018-01-09T09:30:00.55',
                dateEnd: '2018-01-09T11:45:00.55',
                users: [],
              },
              {
                title: 'Событие 3',
                dateStart: '2018-01-09T18:30:00.55',
                dateEnd: '2018-01-09T20:45:00.55',
                users: [],
              }
            ]
          }
        ]
      }
    ];
    shallowExpect(
      <EventsDiagram
        floors={floor}
        classes={['EventDiagram-Story']}
      />)
      .toMatchSnapshot();
  });

  it('should show an events tooltip when an event is clicked', () => {
    const floor = [
      {
        floor: 7,
        rooms: [
          {
            title: 'Ржавый Фред',
            capacity: 7,
            events: [
              {
                title: 'Событие 1',
                dateStart: '2018-01-09T06:30:00.55',
                dateEnd: '2018-01-09T08:45:00.55',
                users: [],
              }
            ]
          }
        ]
      }
    ];
    const wrapper = mount(
      <EventsDiagram
        floors={floor}
        classes={['EventDiagram-Story']}
      />
    );

    wrapper
      .find('.RoomTimeline-Slot_event')
      .simulate('click');

    expect(wrapper.find(EventTooltip)).toHaveLength(1);
  });
});