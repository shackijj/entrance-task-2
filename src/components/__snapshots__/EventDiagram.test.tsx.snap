// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventDiagram should render a floors list 1`] = `
<div
  className="EventsDiagram EventDiagram-Story"
>
  <div
    className="EventsDiagram-HorizontalScrollContainer"
  >
    <Timeline
      classes={
        Array [
          "EventsDiagram-Timeline",
        ]
      }
      hourEnd={23}
      hourStart={7}
    />
    <div
      className="EventsDiagram-VerticalScrollContainer"
    >
      <div
        className="EventDiagram-Floor"
        key="0"
      >
        <div
          className="EventDiagram-FloorTitle"
        >
          7
           этаж
        </div>
        <div
          className="EventDiagram-Room"
          key="0"
        >
          <Room
            capacity={7}
            events={
              Array [
                Object {
                  "dateEnd": "2018-01-09T08:45:00.55",
                  "dateStart": "2018-01-09T06:30:00.55",
                  "title": "Событие 1",
                  "users": Array [],
                },
                Object {
                  "dateEnd": "2018-01-09T11:45:00.55",
                  "dateStart": "2018-01-09T09:30:00.55",
                  "title": "Событие 2",
                  "users": Array [],
                },
                Object {
                  "dateEnd": "2018-01-09T20:45:00.55",
                  "dateStart": "2018-01-09T18:30:00.55",
                  "title": "Событие 3",
                  "users": Array [],
                },
              ]
            }
            title="Ржавый Фред"
          />
          <RoomTimeline
            capacity={7}
            events={
              Array [
                Object {
                  "dateEnd": "2018-01-09T08:45:00.55",
                  "dateStart": "2018-01-09T06:30:00.55",
                  "title": "Событие 1",
                  "users": Array [],
                },
                Object {
                  "dateEnd": "2018-01-09T11:45:00.55",
                  "dateStart": "2018-01-09T09:30:00.55",
                  "title": "Событие 2",
                  "users": Array [],
                },
                Object {
                  "dateEnd": "2018-01-09T20:45:00.55",
                  "dateStart": "2018-01-09T18:30:00.55",
                  "title": "Событие 3",
                  "users": Array [],
                },
              ]
            }
            hourEnd={23}
            hourStart={7}
            title="Ржавый Фред"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
